sdk/include/opentelemetry/sdk/resource/resource.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/trace/attribute_utils.h"
+#include "opentelemetry/sdk/version/version.h"
+#include "opentelemetry/version.h"
+
+#include <cstdlib>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+using ResourceAttributes =
+    std::unordered_map<std::string, opentelemetry::sdk::trace::SpanDataAttributeValue>;
+
+class Resource
+{
+public:
+  Resource(const ResourceAttributes &attributes = ResourceAttributes()) noexcept;
+
+  const ResourceAttributes &GetAttributes() const noexcept;
+
+  std::shared_ptr<Resource> Merge(const Resource &other);
```

Comment:
Why not just return a stack-allocated resource here?  Wouldn't that be a simpler api?

Similar for `Create`.

Reply:
As it is possible to share resources across multiple spans, it would be better to use `shared_ptr` and avoid un-necessary copying.

Reply:
This method doesn't need to determine how resources are shared.

A few points:
- Resource-attachment to Trace is NOT specified in the API.  In fact, I don't think you *can* attach via the API in any other language.
- The resource will be attached to a metric provider or trace provider (see: https://github.com/open-telemetry/opentelemetry-java/blob/efd559f1e09e02f6f5a93ac91c272042849b3bd1/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java#L35) for an example in Java.

I think it's reasonable for *this* method to return on the Stack.   We then have our MetricProvider + TraceProvider have a `Resource` attribute they own, and Trace/Metrics generated from those provides can use Pointers or references to the resource associated with a provider.  

Resources should really *only* be a startup concern and not something we expect users to pass around.  I'd rather see a `Resource&& ` argument on `*Provider` classes just to make ownership clear.   This is similar to other APIs.

References:
- [Trace API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md) has no reference to `Resource`
- [Trace SDK spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md) denotes the resource is *implicitly* associated with a Span.

Reply:
I agree here. `Resource` need to be associated with TracerProvider level, and we can actually pass the reference of same to traces generated from the provider. And in that case, we don't need to manage smart pointers. Let me add other half in this PR, and it would be easier to comprehend.

Reply:
OK,  `Resources::Merge()` and `Resources::Create()` are modified to return stack allocated `Resource` by value. And this resource is then owned by `sdk::tracer_provider` and passed as reference to span. Please refer to `examples/simple/main.cc` in this PR on how the usage would be, and let me know if any issues now.